### PR TITLE
Use cross-compilation where possible

### PR DESCRIPTION
**Pull request by victorlin**
### Description of proposed changes

As [noted](https://github.com/nextstrain/docker-base/pull/50#discussion_r986331019) in the PR that introduced multi-platform images, [cross-compilation](https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/) can substantially improve multi-platform build time.

### Related issue(s)

- #50

### Notable discussions

- Naming of Dockerfile stages ([ref](https://github.com/nextstrain/docker-base/pull/98#discussion_r1133001444))

### Testing

- [x] Checks pass
- [x] Verify that binaries on the `arm64` image variant are native to the platform. Evidence:

    ```
    % docker pull --platform linux/arm64 nextstrain/base:branch-victorlin-arm64-cross-compilation
    % docker run -it --entrypoint /bin/bash nextstrain/base:branch-victorlin-arm64-cross-compilation

    nextstrain:~/build $ apt-get update && apt-get install file -y

    nextstrain:~/build $ file /usr/local/bin/raxmlHPC-PTHREADS-AVX
    /usr/local/bin/raxmlHPC-PTHREADS-AVX: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=a06a245ef61d80661f3c3fd584a4a2002a4568f3, for GNU/Linux 3.7.0, not stripped

    nextstrain:~/build $ file /usr/local/bin/FastTreeDblMP
    /usr/local/bin/FastTreeDblMP: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=4c9df1efc0725502472822181484324f8aa3be59, for GNU/Linux 3.7.0, not stripped

    nextstrain:~/build $ file /usr/local/bin/vcftools
    /usr/local/bin/vcftools: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=e9b3149b8288e62c72d82f420d423076f22cf490, for GNU/Linux 3.7.0, with debug_info, not stripped

    nextstrain:~/build $ node -e "console.log(process.arch)"
    arm64
    ```

### CODE REVIEW

File: Dockerfile
Comment (tsibley) [inline Dockerfile:5]: I think this description needs some rewording and clarification. And in doing so, we might want to rename the builder images.

There's two intersecting things here, right?

1. build vs. target platform
2. if build == target or not

Names and commentary seem to assume that build != target?

But when build == target, then emulation is never in use and `native-builder` and `emulated-builder` don't make much sense as names? I think the use of "emulated" here is assuming the scenario of a Linux x86-64 build platform targeting aarch64.

**Review by tsibley — COMMENTED** [review]
On my first read thru this broadly makes sense to me. I left a few comments, the most noteworthy about clarification of "native" vs. "emulated" in the context of the build platform == target platform.

I haven't tried anything out yet but will plan to do so next week.

File: Dockerfile
Comment (victorlin) [reply Dockerfile:99]: I forgot to update this commit. I'll move the repo to @nextstrain and download as a tarball.

File: Dockerfile
Comment (victorlin) [reply Dockerfile:5]: Yes thanks for breaking that down, I'll reword. As for names, `strictly-native-builder` and `potentially-emulated-builder` would be the most accurate, but are quite verbose... any other suggestions?

File: Dockerfile
Comment (victorlin) [reply Dockerfile]: Done: 623fd780afbf22901d1f576117deda4a0cb84189

File: Dockerfile
Comment (victorlin) [reply Dockerfile]: [Done in force-push](https://github.com/nextstrain/docker-base/compare/53cd37bebb69b25ec5b086f823ee072989cf8b35..90c199a265138d5af3fb115f4a95dacc7b0d9f06#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R224).

File: Dockerfile
Comment (tsibley) [reply Dockerfile:5]: maybe

```
build-platform-builder
target-platform-builder
```

?

Comment (victorlin) [reply Dockerfile]: Reworded in [force-pushes](https://github.com/nextstrain/docker-base/compare/90c199a265138d5af3fb115f4a95dacc7b0d9f06..5478afcd2b3cfcbf72568b96be18f05dd8ecb643).

I opted for 

```
builder-build-platform
builder-target-platform
```

so the "builder" part of the image name wouldn't be so separated from "base" while keeping similar image and stage names (`nextstrain/base-build-platform-builder` vs. `nextstrain/base-builder-build-platform`). Still a bit verbose and "builder-build-platform" is a bit confusing, but I don't have anything better.

Comment (victorlin) [reply Dockerfile]: @tsibley Once we settle on a good set of names here, I'll need you to create the respective repositories in the Nextstrain Docker Hub organization (and delete the old names) so the images can be pushed there.

File: Dockerfile
Comment (tsibley) [reply Dockerfile:27]: I guess we can remove g++ since iltorb is not required for Auspice as of the next release?

File: Dockerfile
Comment (tsibley) [reply Dockerfile]: > Still a bit verbose and "builder-build-platform" is a bit confusing, but I don't have anything better.

Seems good enough to me, as far as warts go, since they're really only visible/matter to us.

I'll create the repos. (Didn't realize you couldn't!)

Comment (tsibley) [reply Dockerfile]: https://hub.docker.com/u/nextstrain/base-builder-build-platform
https://hub.docker.com/u/nextstrain/base-builder-target-platform

File: Dockerfile
Comment (victorlin) [reply Dockerfile:27]: Yes, I'll revert 373ebcc55438ff2084dd8e2661e126bb5f5a00a5 after the next Auspice release.

**Review by tsibley — APPROVED** [review]
On the whole, this looks good to me!

I haven't closely tested the final image produced, but trust your testing and CI here (esp. given recent CI improvements).

File: Dockerfile
Comment (victorlin) [reply Dockerfile]: Oh, splitting the comment away from c3fbbd9bbe09edb6c14b5cf3c1ad8c56ee2a9aa9 confused me. I've combined the two in 7e7bfb29149e991c74c30c6c7e22e37a9db84fc2 and reworded the comment. I still don't fully understand the implications of `npm install auspice` with a `linux/amd64` image and copying those files onto the `linux/arm64` image, so I'll test more on my M1 Mac.

File: devel/build
Comment (victorlin) [reply devel/build]: No good reason. Updated the variable names in 0b154146d2ec0f0ebfbf8a0d5768e321632a1ea4.

Comment (tsibley) [reply devel/build]: :pray: Ah, but looks like devel/copy-images got one variable renamed (target platform) but not the other (build platform)…

Comment (victorlin) [reply devel/build]: 🤦 good catch, fixed in [force-push](https://github.com/nextstrain/docker-base/compare/fcfc6c76c69a21fc4a42ed13a2d06141796473fe..43f57d7bc7730a0d5e6659e478ffeb3b5b803373).

**Review by tsibley — APPROVED** [review]
This seems good to go to me once CI passes.

I'll be curious to see build times before/after…

File: Dockerfile
Comment (tsibley) [reply Dockerfile]: > I still don't fully understand the implications of `npm install auspice` with a `linux/amd64` image and copying those files onto the `linux/arm64` image, so I'll test more on my M1 Mac.

I would think (maybe naively) that at worst this could result in Node.js extensions running under emulation. But, the `--arch` and `--target_arch` flags presumably avoid that?

So (empirically-driven understanding)++

Comment (victorlin) [reply Dockerfile]: That's also the worst case I can think of. The command runs without errors, so at least *something* must have installed. A summary of my understanding:

1. The flags are passed down from `npm` to any precompiling tools that are used.
2. It's up to each dependency to support cross-platform installs through one of those methods (see all the PRs/issues referencing https://github.com/nodejs/node-gyp/issues/829).
3. If a dependency does not support installing for the target platform through `node-gyp`, the calling NPM command will error. I confirmed this by running `npm install --arch=bogus electron` which gives a 404 error.
4. If a dependency does not support installing for the target platform through one of the other precompiling tools, I'm not sure what happens but I assume either:
    1. The calling `npm` command will error, as is the case with `node-gyp`.
    2. Fallback to the host architecture.

Looking at the bigger picture, I think it's fine to use emulation for Auspice on `linux/arm64` hosts.

Comment (victorlin) [reply Dockerfile]: I'm tempted to split this out into a separate PR since `--arch` and `--target_arch` don't seem to be used for any dependencies currently. I can't find a good method to tell if that's the case, but worst case Auspice runs with emulation on `linux/arm64` hosts.

My plan:

1. Replace 7e7bfb29149e991c74c30c6c7e22e37a9db84fc2 with a comment on potential emulation of dependencies.
2. Wait for CI to build a new test image.
3. Verify that Auspice works locally on my M1 machine using the test image.

Comment (tsibley) [reply Dockerfile]: Sounds fine if you want to split it out, but I also think ok to leave it here.

Comment (victorlin) [reply Dockerfile]: I came up with a search pattern for platform-specific dependencies, and did not find anything that sticks out as impactful (`node-notifier` and `puppeteer` would only be used for testing purposes). I'll follow through with the plan.

```
nextstrain:~/auspice $ pwd
/nextstrain/auspice

nextstrain:~/auspice $ find node_modules -executable -type f -exec file {} + | grep 64
node_modules/node-notifier/vendor/mac.noindex/terminal-notifier.app/Contents/MacOS/terminal-notifier:                        Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>
node_modules/node-notifier/vendor/notifu/notifu64.exe:                                                                       PE32+ executable (GUI) x86-64, for MS Windows
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/nacl_irt_x86_64.nexe:                                       ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=7aeb4f45ea5cec7d8e4184264ad39f0f77bcaee2, stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/nacl_helper_bootstrap:                                      ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=5c5f3935c8f8a15ba325f0d73dfa585fa9390cf9, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/ClearKeyCdm/_platform_specific/linux_x64/libclearkeycdm.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/libEGL.so:                                                  ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/crashpad_handler:                                           ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/chrome_sandbox:                                             ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/chrome:                                                     ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/libGLESv2.so:                                               ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/swiftshader/libEGL.so:                                      ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/swiftshader/libGLESv2.so:                                   ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, not stripped
node_modules/puppeteer/.local-chromium/linux-722234/chrome-linux/nacl_helper:                                                ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, not stripped
```

Comment (tsibley) [reply Dockerfile]: Oh jeez, we're shipping dev deps in the container?! Including a copy of Chromium! I guess we are since NPM installs them by default. Oy. We should (separate from this PR) be using `--omit dev` or equivalent. Looks like that'll roughly reduce uncompressed image size by 600MB.

Comment (victorlin) [reply Dockerfile]: Agreed! Glad the extra digging here was worth it.

Comment (victorlin) [reply Dockerfile]: Replaced 7e7bfb29149e991c74c30c6c7e22e37a9db84fc2 with deabcc1a6fa34a75a0b9e4b6b8440ca04e71962d.

Comment (tsibley) [reply Dockerfile]: https://github.com/nextstrain/docker-base/pull/165

Comment (victorlin) [reply Dockerfile]: Auspice works on my M1 machine using the latest CI image:

```
nextstrain view --docker --image nextstrain/base:branch-victorlin-arm64-cross-compilation zika-tutorial/auspice/
```

Merging!

**Pull request merged**
